Fixes issue #814 Race condition to ObservablePropertyHelper

bradphelan opened the pull request:
Using multicast I think was the wrong solution as there was no real need to multicast. A simpler solution which I have used in the past is to create a BehaviourSubject<IObservable<T>>() initialized with an Observable.Never<T>(). 

I've created a test case to show the previous failure and it now passes with the changed code.

bradphelan commented:
Fixes #814

bradphelan commented:
@paulcbetts Any chance to have this integrated. It's got test cases and all.

jlaanstra commented:
I am not entirely convinced yet that we should take this change. The actual fix seems to be the _activated flag in the getter of Value. The rest is just refactoring of code.

To me it seems weird that to setup property A, you somehow need the value of property A.

bradphelan commented:
@jlaanstra It's not really wierd. It's just lazy initialization. It avoids subscription to the underlying observable unless the value is actually ever used. It's possibly an over optimization but that is what the original code was attempting to do but failed to implement correctly. Using an initial sequence of Observable.Never and switching it to the real sequence on init is more robust than the previous code.

bradphelan commented:
@paulcbetts @jlaanstra Guys! I have a pull request here with a test case showing a verified bug and that the code fixes the error but I'm getting no feedback. What can I do to get this pull request accepted?

gluck commented:
Not having any authority here, but here's my 2 cents advice about this change.
While the issue is real (reentrancy on `ObservableAsPropertyHelper::get_Value` causes duplicate subscription), I agree with @jlaanstra that the proposed fix goes beyond the scope.

I'll suggest some edits in comments.

gluck commented on ReactiveUI.Tests/ObservableAsPropertyHelperTest.cs:
```diff
@@ -132,4 +138,79 @@ public void ToPropertyShouldFireBothChangingAndChanged()
+
+
+                var count = 0;
+                var obs = Observable.Create<bool>(observer =>
```
can be simply:
```
     Observable.Return(true).Do(_ => count++);
```

gluck commented on ReactiveUI.Tests/ObservableAsPropertyHelperTest.cs:
```diff
@@ -132,4 +138,79 @@ public void ToPropertyShouldFireBothChangingAndChanged()
+
+                var f = new Fixture();
+
+                f._A = obs.ToProperty(f, t => t.A, false);
```
Maybe prefer using the recommended way of declaring properties ? i.e. move it to the VM class and use:
```
Observable.Return(true).Do(_ => _count++).ToProperty(this, x => x.A, out _A);
```

gluck commented on ReactiveUI.Tests/ObservableAsPropertyHelperTest.cs:
```diff
@@ -132,4 +138,79 @@ public void ToPropertyShouldFireBothChangingAndChanged()
+                // or RX bug.
+                f.PropertyChanged += ( e, s ) => Console.WriteLine(f.A);
+
+                var v = f.A;
```
Add an AssertEquals to avoid the unused variable ?

gluck commented on ReactiveUI/ObservableAsPropertyHelper.cs:
```diff
@@ -112,10 +114,17 @@ public sealed class ObservableAsPropertyHelper<T> : IHandleObservableErrors, IDi
 
+        private void Activate()
+        {
+            if ( _activated )
```
I would use Interlock together with a SingleAssignmentDisposable to guarantee both single subscription and disposal, keeping the previous logic unchanged:
```
                if (Interlocked.CompareExchange(ref _activated, 1, 0) == 0) {
                    _disp.Disposable = _source.Connect();
                }
```

bradphelan commented:
Thanks for the feedback. I'll make some updates and resubmit

bradphelan commented:
Thanks @gluck for the feedback. I've reverted the changes and applied the fix only to the Value.get as you suggested. I've also cleaned up the test case as per your suggestions. I've checked the test case still fails without the fixes applied and that it passes with the fixes applied.

gluck commented on ReactiveUI.Tests/ObservableAsPropertyHelperTest.cs:
```diff
@@ -132,4 +138,78 @@ public void ToPropertyShouldFireBothChangingAndChanged()
         }
     }
+
+    class UnitTestMode : IModeDetector
```
pretty minor but this could be a IDisposable helper so one'd write the test like that (so that all test scope is within the test method):

    using (new ProductionMode()) {
        ...
    }

also minor but I believe you can use new PlatformModeDetector() when resetting the mode (it's the default).

bradphelan commented:
@gluck Changes on your latest suggestions have been made. We would really appreciate this pull request being merged with the next release. We see side effects of this bug throughout our application.

gluck commented on RxUI.DotSettings:
```diff
@@ -1,18 +1,7 @@
-	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/UserRules/=9b2b4c09_002Dbe0f_002D4b8e_002D8f1a_002Dcdccf074d45c/@EntryIndexedValue">&lt;Policy&gt;&lt;Descriptor Staticness="Static, Instance" AccessRightKinds="Private" Description="Private Fields/Methods/Properties"&gt;&lt;ElementKinds&gt;&lt;Kind Name="FIELD" /&gt;&lt;Kind Name="READONLY_FIELD" /&gt;&lt;Kind Name="CONSTANT_FIELD" /&gt;&l ...
-	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateBlankLinesAroundFieldToBlankLinesAroundProperty/@EntryIndexedValue">True</s:Boolean></wpf:ResourceDictionary>
\ No newline at end of file
+<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
change made on purpose ?

bradphelan replied:
I reverted it in the next commit.

gluck replied:
seems to me that you reverted more than your original modif, I still see this file changed on the overall diff

bradphelan commented:
@gluck  It's reverted correctly now.

gluck commented:
LGTM, now we need to get a :shipit: !

eggapauli commented:
@kentcb Can we merge this to reactive-ui. This bug is killing me in production!

kentcb commented:
@eggapauli : it's targeting the wrong branch (needs to be rxui7-master). Any chance @bradphelan could re-target?
